### PR TITLE
8311639: Replace currentTimeMillis() with nanoTime() in jtreg/gc

### PR DESCRIPTION
**Pull request by lkorinth**
I have removed usage of `currentTimeMillis()` in jtreg/gc as `currentTimeMillis()` is not monotonic. It is mostly changing from milliseconds to nanoseconds. In certain places, I have changed the code from something like `(instant1 < instant2)` to `(instant1 - instant2 < 0)` It might look like I am removing instant2 from both sides of the inequality and the result ought to be the same, but due to overflow arithmetic, the change should be better if nanoseconds where to overflow. I have also removed some loops where the loop is doing nothing except sleeping. I somewhat shortened the string in the key of a hash map because the nanosecond value made the string too long for the heap size.

All tests within jtreg/gc passes.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311639](https://bugs.openjdk.org/browse/JDK-8311639): Replace currentTimeMillis() with nanoTime() in jtreg/gc (**Enhancement** - P4)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15331/head:pull/15331` \
`$ git checkout pull/15331`

Update a local copy of the PR: \
`$ git checkout pull/15331` \
`$ git pull https://git.openjdk.org/jdk.git pull/15331/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15331`

View PR using the GUI difftool: \
`$ git pr show -t 15331`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15331.diff">https://git.openjdk.org/jdk/pull/15331.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15331#issuecomment-1682548865)

### CODE REVIEW

File: test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java [hunk 1]
```diff
@@ -77,7 +77,7 @@ class LongLivedAllocationTask extends Exitable implements Runnable {
     @Override
     public void run() {
         while (!shouldExit()) {
-            String prefix = "long" + System.currentTimeMillis();
+            String prefix = "long" + (System.nanoTime() % 10000); // limit to 4 digits after changing from milliseconds to nanoseconds, else the key will use more memory
```
Comment (albertnetymk) [inline test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java:80]: Is this used to calculate elapsed time? If not, I think using `currentTimeMillis` is fine.

File: test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java [hunk 2]
```diff
@@ -191,9 +191,9 @@ public static void main(String[] args) throws Exception {
         if (args.length == 0) {
             throw new IllegalArgumentException("Must specify timeout in seconds as first argument");
         }
-        int timeout = Integer.parseInt(args[0]) * 1000;
-        System.out.println("Running with timeout of " + timeout + "ms");
-        endTime = System.currentTimeMillis() + timeout;
+        long timeoutNanos = Integer.parseInt(args[0]) * 1_000_000_000L;
+        System.out.println("Running with timeout of " + timeoutNanos + "ns");
```
Comment (stefank) [inline test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java:195]: The precision here is in seconds. Maybe just leave this print out to be in ms?

**Review by stefank — APPROVED** [review]
Looks good. I'd also go with Albert's suggestion and keep timeout logging in TestSystemGC.

File: test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java (hunk 1 above)
Comment (lkorinth) [reply test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java:80]: It is not to calculate elapsed time. It is to create a unique(ish) key for a hash table. The new code does not guarantee a unique value, but then neither did the old.

Comment (lkorinth) [reply test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java:80]: I could possibly add some characters to the key if that is what bothers you. I could of course also use `currentTimeMillis` as it is nothing wrong with it. But I would prefer removing `currentTimeMillis` in all gc testing.

Comment (lkorinth) [reply test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java:80]: I will try to see if I can do it without using time at all.

File: test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java (hunk 2 above)
Comment (lkorinth) [reply test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java:195]: I will change it to seconds.

File: test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java (hunk 1 above)
Comment (albertnetymk) [reply test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java:80]: Not using time is probably the best. Not obvious if #digits here is critical. Maybe this is useful -- e.g. to generate fixed 6-digit int:

```java
Random rnd = new Random();
int n = 100_000 + rnd.nextInt(900_000);
```

**Comment by lkorinth** [comment]
I will refactor a bit more...

**Review by albertnetymk — APPROVED** [review]

File: test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java (hunk 1 above)
Comment (lkorinth) [reply test/hotspot/jtreg/gc/stress/systemgc/TestSystemGC.java]: I reverted to the original implementation as I did not understand the idea behind it.

**Review by stefank — APPROVED** [review]

**Comment by lkorinth** [comment]
Thanks Albert and StefanK, I will integrate next week.

**Pull request closed without merging**